algorithmic_efficiency/checkpoint_utils.py

```diff
@@ -0,0 +1,162 @@
+
+  # Load_latest_checkpoint() will return checkpoint_state if
+  # checkpoint_dir does not exist or if it exists and contains no checkpoints.
+  found_checkpoint = (latest_ckpt['global_step'] != uninitialized_global_step)
+
+  if not found_checkpoint:
+    return (optimizer_state,
+            model_params,
+            model_state,
+            train_state,
+            eval_results,
+            global_step,
+            preemption_count)
+
+  # If there's the latest checkpoint in the checkpoint_dir, restore from that.
+  if framework == 'jax':
+    checkpoint_state = replicate_checkpoint(
+        latest_ckpt,
+        pytree_keys=[
+            'optimizer_state',
+            'model_params',
+            'model_state',
+        ])
+    checkpoint_state['optimizer_state'] = (checkpoint_state['optimizer_state'],
+                                           opt_update_fn)
+    checkpoint_state['eval_results'] = [
+        (value, key) for key, value in latest_ckpt['eval_results'].items()
+    ]
+
+  else:
+    checkpoint_state = latest_ckpt
+    model_params.load_state_dict(checkpoint_state['model_params'])
+    checkpoint_state['model_params'] = model_params
+    for key in optimizer_state.keys():
+      optimizer_state[key].load_state_dict(
+          checkpoint_state['optimizer_state'][key])
+      checkpoint_state['optimizer_state'][key] = optimizer_state[key]
+
+    logging.info('Loaded checkpoint from %s', save_path)
+  return (checkpoint_state['optimizer_state'],
+          checkpoint_state['model_params'],
+          checkpoint_state['model_state'],
+          checkpoint_state['train_state'],
+          list(checkpoint_state['eval_results']),
+          checkpoint_state['global_step'],
+          checkpoint_state['preemption_count'] + 1)
+
+
+def replicate_checkpoint(latest: dict,
+                         pytree_keys: Sequence[str],
+                         replicate: bool = True) -> dict:
+  pytree = {k: latest[k] for k in pytree_keys}
+  if replicate:
+    pytree = jax_utils.replicate(pytree)
+  extra_dict = {k: latest[k] for k in latest.keys() if k not in pytree_keys}
+  pytree.update(extra_dict)
+  return pytree
+
+
+def save_checkpoint(framework: str,
+                    optimizer_state: spec.OptimizerState,
+                    model_params: spec.ParameterContainer,
+                    model_state: spec.ModelAuxiliaryState,
+                    train_state: dict,
+                    eval_results: list,
+                    global_step: int,
+                    preemption_count: int,
+                    checkpoint_dir: str) -> None:
+  if RANK == 0:
+    if framework == 'jax':
+      model_params = jax.device_get(jax_utils.unreplicate(model_params))
+      opt_state, _ = optimizer_state
+      opt_state = jax.device_get(jax_utils.unreplicate(opt_state))
+      model_state = jax.device_get(jax_utils.unreplicate(model_state))
+    else:
+      model_params = model_params.state_dict()
```

Comment:
To also support DP instead of only DDP, I think we can just add these two lines above this one:
```python
if isinstance(model_params, torch.nn.DataParallel):
  model_params = model_params.module
```

Reply:
Thank you! Both with and without did not cause problems, but I did find some threads recommending to `.module` for DataParallel models. I updated the code to do this both when saving and loading.

Reply:
Maybe it works if you also load the state dict to a model wrapped with DataParallel, but if not it would fail?

Reply:
Ah, that is a good point. Yes, I think that might be the case, since `nn.DataParallel` parent is `nn.Module`, so it should have functionality the load and save state_dict.